crates/legacy/src/builder_state.rs

```diff
@@ -1143,3 +1141,414 @@ impl<TYPES: NodeType> BuilderState<TYPES> {
+        async_compatibility_layer::logging::setup_backtrace();
+        tracing::info!("Testing the builder core with multiple messages from the channels");
+
+        // Number of views to simulate
+        const NUM_ROUNDS: usize = 5;
+        // Number of transactions to submit per round
+        const NUM_TXNS_PER_ROUND: usize = 4;
+        // Capacity of broadcast channels
+        const CHANNEL_CAPACITY: usize = NUM_ROUNDS * 5;
+        // Number of nodes on DA committee
+        const NUM_STORAGE_NODES: usize = 4;
+
+        // start builder_state without entering event loop
+        let (_senders, global_state, mut builder_state) =
+            start_builder_state_without_event_loop(CHANNEL_CAPACITY, NUM_STORAGE_NODES).await;
+
+        // insert some builder states
+
+        // Transactions to send
+        let all_transactions = (0..NUM_ROUNDS)
+            .map(|round| {
+                (0..NUM_TXNS_PER_ROUND)
+                    .map(|tx_num| TestTransaction::new(vec![round as u8, tx_num as u8]))
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
+        let mut prev_quorum_proposal: Option<QuorumProposal<TestTypes>> = None;
+        #[allow(clippy::needless_range_loop)]
+        for round in 0..NUM_ROUNDS {
+            let transactions = all_transactions[round].clone();
+            let (quorum_proposal, _quorum_proposal_msg, _da_proposal_msg, builder_state_id) =
+                calc_proposal_msg(NUM_STORAGE_NODES, round, prev_quorum_proposal, transactions)
+                    .await;
+            prev_quorum_proposal = Some(quorum_proposal.clone());
+            let (req_sender, _req_receiver) = broadcast(CHANNEL_CAPACITY);
+            let leaf: Leaf<TestTypes> = Leaf::from_quorum_proposal(&quorum_proposal);
+            let leaf_commit = RawCommitmentBuilder::new("leaf commitment")
+                .u64_field("view number", leaf.view_number().u64())
+                .u64_field("block number", leaf.height())
+                .field("parent Leaf commitment", leaf.parent_commitment())
+                .var_size_field(
+                    "block payload commitment",
+                    leaf.payload_commitment().as_ref(),
+                )
+                .finalize();
+            global_state.write_arc().await.register_builder_state(
+                builder_state_id,
+                ParentBlockReferences {
+                    view_number: quorum_proposal.view_number,
+                    vid_commitment: quorum_proposal.block_header.payload_commitment,
+                    leaf_commit,
+                    builder_commitment: quorum_proposal.block_header.builder_commitment,
+                },
+                req_sender,
+            );
+        }
+
+        // send out a decide event
+        // randomly choose a latest_decide_view_number between [0, NUM_ROUNDS]
+        let latest_decide_view_number = ViewNumber::new(3);
```

Comment:
![image](https://github.com/user-attachments/assets/a44800b0-9038-4784-a503-f2c11ee4b69b)

I think the comment is a bit misleading here 😄

Reply:
[18befc0](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/18befc025d8ee08aa023f33e4301b464366095b8)